projects/baby_driver/scripts/gpio_interrupts.py

```diff
@@ -0,0 +1,182 @@
+"""Python implementation of register_gpio_interrupt function"""
+import can_util
+from gpio_port import GpioPort
+import message_defs
+
+NUM_PINS_PER_PORT = 16
+
+# pylint: disable=too-few-public-methods
+class InterruptEdge:
+    """Stores interrupt edge values"""
+    RISING = 0
+    FALLING = 1
+    RISING_AND_FALLING = 2
+    NUM_INTERRUPT_EDGES = 3
+
+# Callbacks that user defines when registering interrupt are stored here
+callback_dict = {
+    #("port", "pin"): callback,
+}
+
+def callback_listener(can_message):
+    """
+    Called by the notifier everytime a message is received from the firmware side notifiying
+    of an interrupt. It filters through the messages to find the can message with babydriver
+    message ID GPIO_IT_INTERRUPT and calls user-defined/default callback that is registered
+    with the interrupt
+    """
+
+    unfiltered_msg = can_util.Message.from_msg(can_message)
+    # print(f"{unfiltered_msg.message_id}, {unfiltered_msg.data[0]},
+    # {unfiltered_msg.data[1]}, {unfiltered_msg.data[1]}")
+    if (unfiltered_msg.message_id == message_defs.BABYDRIVER_CAN_MESSAGE_ID and
+        unfiltered_msg.data[0] == message_defs.BabydriverMessageId.GPIO_IT_INTERRUPT):
+
+        data = unfiltered_msg.data[:]
+        port = data[1]
+        pin = data[2]
+        edge = data[3]
+        # Calling function from callback_dict based on (port, pin)
+        try:
+            # ret is only used for testing purposes
+            ret = callback_dict[(port, pin)]((port, pin, edge))
+        except TypeError as type_error:
+            raise "Callback function parameters are of incorrect format (use a named-tuple called \
+                    info to store port, pin and edge" from type_error
+
+
+    return ret
+
+
+def default_callback(info):
+    """ The function called after occurence of interrupt if
+       no user-defined callback function exists"""
+    port, pin, edge = info
+    port = chr(port + ord('A'))
+    print(f"Interrupt on P{port}{pin}: {edge}")
+
+
+#  Setting up notifier to listen for all CAN messages
+def init_notifier_gpio_it():
+    """ Imports notifier from can_util and adds callback_listener
+        to list of listeners the notifier calls upon receiving a can message """
+    notifier = can_util.get_bus_notifier()
+    notifier.add_listener(callback_listener)
+
+
+def register_gpio_interrupt(port, pin, edge = InterruptEdge.RISING, callback = None):
```

Comment:
I think `RISING_AND_FALLING` should be the default edge, since it gives the most information (and it's also the slowest to type).